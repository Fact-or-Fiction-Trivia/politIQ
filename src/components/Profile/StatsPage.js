import React from 'react';
import Button from '@material-ui/core/Button';
import UserScoreboard from '../Leaderboard/UserScoreboard';

const StatsPage = (props) => {
    return (
        <>
            {/* <h1 id="settings-heading">Quiz Stats</h1> */}
            <UserScoreboard uid={props.uid} moneyWon={props.moneyWon} />
            <Button color="primary" variant="contained" id="cashOut">Cash Out</Button>
        </>
    )
}

export default StatsPage;