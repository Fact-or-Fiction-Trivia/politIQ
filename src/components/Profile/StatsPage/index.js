import React, { Component } from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

import Button from '@material-ui/core/Button';
import Modal from '@material-ui/core/Modal';
import Paper from '@material-ui/core/Paper';
import Close from '@material-ui/icons/Close';
import UserScoreboard from '../../Leaderboard/UserScoreboard';

import { db } from '../../../firebase';

class StatsPage extends Component {
  constructor(props) {
    super(props);
    this.state = {
      modalOpen: false,
      cashoutRequested: false,
    };
  }

  componentDidMount() {
    if (this.props.userInfo.cashoutRequested !== null) {
      // user cannot cash out if they have already cashed out or if they have not won any money
      const { cashoutRequested, moneyWon } = this.props.userInfo;
      let requested;
      if (cashoutRequested === "true" || moneyWon === 0) {
        requested = true;
        this.setState({
          cashoutRequested: requested,
        })
      }
    }
  }

  toggleModal = () => {
    this.setState({
      modalOpen: !this.state.modalOpen,
    })
  }

  requestCashOut = () => {
    // send object with uid, email address, moneywon, and date of cash out request
    // update user information object to say that a cashout has been requested
    const { email, moneyWon, displayName } = this.props.userInfo
    const date = moment().format('YYYY-MM-DD')
    const uid = this.props.uid
    db.requestCashOut(uid, date, email, displayName, moneyWon)
        
    const localStorageUser = JSON.parse(localStorage.getItem('authUser'));
    localStorageUser.cashoutRequested = true;
    localStorage.setItem('authUser', JSON.stringify(localStorageUser));
        
    this.setState({
      cashoutRequested: true,
      modalOpen: false,
    })
  }

  render() {
    return (
      <>
        <UserScoreboard uid={this.props.uid} moneyWon={this.props.userInfo.moneyWon} />
        {this.props.userInfo.moneyWon === 0 ? <p id="cashout-instructions">You haven't been awarded any cash prizes yet! Keep playing for your chance to top the leaderboards!</p> : null}
        <Button color="primary" variant="contained" id="cashOut" onClick={this.toggleModal} disabled={this.state.cashoutRequested}>{this.state.cashoutRequested && this.props.userInfo.moneyWon !== 0 ? 'Cash Out Requested' : 'Cash Out' }</Button>
        <Modal
          aria-labelledby="Cash Out"
          aria-describedby="Coming soon - cash out and receive your earnings!"
          open={this.state.modalOpen}
          onClose={this.toggleModal}
          className="cashout-modal"
        >
          <Paper>
            <Close style={{ float: 'right', padding: '1vh', display: 'block'}} onClick={this.toggleModal}/>
            <h3>Coming soon!</h3>
            <p>You have ${this.props.userInfo.moneyWon} available.</p>
            <Button variant="contained" color="primary" style={{ marginBottom: '5vh' }} onClick={this.requestCashOut} id="cashout-confirm">Click Here to Cash Out</Button>
          </Paper>
        </Modal>
      </>
    )
  }
}

StatsPage.propTypes = {
  uid: PropTypes.string.isRequired,
  userInfo: PropTypes.object.isRequired,
}

export default StatsPage;