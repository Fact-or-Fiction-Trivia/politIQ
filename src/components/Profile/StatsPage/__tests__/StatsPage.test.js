import React from 'react';
import { mount } from 'enzyme';

import StatsPage from '../index';

const uid = "GvM7fEkYp4fcaSuScwixV0xinfT2"
const fakeUserInfo = {
    affiliation: "Republican",
    bio: "Forever thinking about chicken nuggs ",
    cashoutRequested: false,
    displayName: "Hannah",
    email: "hannah.werman@gmail.com",
    emailVerified: false,
    invisibleScore: false,
    isAdmin: false,
    lastActive: "2019-08-06",
    lifetimeEarnings: 5,
    moneyWon: 5,
    soundsOn: true,
    uid: "GvM7fEkYp4fcaSuScwixV0xinfT2"
}

describe('<StatsPage />', () => {
    it('should render correctly and match snapshot', () => {
        const wrapper = mount(<StatsPage uid={uid} userInfo={fakeUserInfo} />);
        
        expect(wrapper).toMatchSnapshot();

        const button = wrapper.find('button#cashOut')
        expect(button.text()).toEqual('Cash Out')
    })

    it('button click should toggle a modal if the user has not previously requested a cashout and has a nonzero moneyWon', async () => {
        global.localStorage.setItem('authUser', JSON.stringify(fakeUserInfo));

        const wrapper = mount(<StatsPage uid={uid} userInfo={fakeUserInfo} />);

        const button = wrapper.find('button#cashOut')
        expect(wrapper.state('modalOpen')).toBe(false)
        expect(button.props().disabled).toBe(false)
        button.simulate('click')
        expect(wrapper.state('modalOpen')).toBe(true)

        // expect a modal component to be rendered
        wrapper.update()
        const modal = wrapper.find('Modal')

        expect(modal.props().open).toBe(true)

        global.localStorage.clear();
    })

    it('button click should be disabled if user has requested a cashout', async () => {
        const wrapper = mount(<StatsPage uid={uid} userInfo={fakeUserInfo} />);

        wrapper.setState({ cashoutRequested: true });
        wrapper.update();

        expect(wrapper.state().cashoutRequested).toBe(true);
        const button = wrapper.find('button#cashOut');
        expect(button.props().disabled).toBe(true)
        button.simulate('click');
        // button click should not trigger the modal because the button is disabled
        expect(wrapper.state('modalOpen')).toBe(false)
    })
})

