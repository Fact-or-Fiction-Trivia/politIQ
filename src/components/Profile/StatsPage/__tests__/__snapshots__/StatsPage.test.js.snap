// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StatsPage /> should render correctly and match snapshot 1`] = `
<Fragment>
  <Component
    moneyWon={5}
    uid="GvM7fEkYp4fcaSuScwixV0xinfT2"
  />
  <CashOutButton
    uid="GvM7fEkYp4fcaSuScwixV0xinfT2"
    userInfo={
      Object {
        "affiliation": "Republican",
        "bio": "Forever thinking about chicken nuggs ",
        "cashoutRequested": false,
        "displayName": "Hannah",
        "email": "hannah.werman@gmail.com",
        "emailVerified": false,
        "invisibleScore": false,
        "isAdmin": false,
        "lastActive": "2019-08-06",
        "lifetimeEarnings": 5,
        "moneyWon": 5,
        "soundsOn": true,
        "uid": "GvM7fEkYp4fcaSuScwixV0xinfT2",
      }
    }
  />
</Fragment>
`;
